File: C++/题目1444：More is better.cpp

#include <stdio.h>
const int maxm = 10000002;
int parent[maxm];
int num[maxm];
int n;
int i;
 
int findParent(int f) {  
    while(parent[f] != f){
        f = parent[f];
    }
    return f;
}  
 
void unionTwo(int f, int t) {  
             
    
}  
 
int max(int a, int b){
	return a > b ? a : b;
}

int main(){
    while(scanf("%d",&n) != EOF){
        for(i = 1; i < maxm; i++){
			parent[i] = i;
			num[i] = 1;
        }
		int maxNum = 1;
        for(i = 0 ; i < n ; i++){
			int a, b;
            scanf("%d%d",&a,&b);
            a = findParent(a);  
			b = findParent(b);  
			if (a > b) {     
				parent[a] = b;
				num[b] += num[a];
				if(num[b] > maxNum){
					maxNum = num[b];
				}
			}else if(a < b){     
				parent[b] = a;
				num[a] += num[b];
				if(num[a] > maxNum){
					maxNum = num[a];
				}
			}	  
        }
		
        printf("%d\n",maxNum);
        
    }
    return 0;
}

/**************************************************************
	Problem: 1444
	User: wangzhenqing
	Language: C++
	Result: Accepted
	Time:600 ms
	Memory:79144 kb
****************************************************************/


                        

File: Java/题目1444：More is better.java

// 题目1444：More is better

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:55
 * @url:http://ac.jobdu.com/problem.php?pid=1444
 * 解题思路参考csdn:http://blog.csdn.net/u013027996/article/details/17174509
 */

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;

public class Main {
    /*
     * 1444
     */
    private static int[] parent = new int[10000001];

    private static int[] num = new int[10000001];

    public static void main(String[] args) throws Exception {
        StreamTokenizer st = new StreamTokenizer(new BufferedReader(
                new InputStreamReader(System.in)));
        while (st.nextToken() != StreamTokenizer.TT_EOF) {
            int n = (int) st.nval;
            if (n == 1) {
                System.out.println(2);
            } else {
                for (int i = 1; i < 10000000; i++) {
                    parent[i] = i;
                    num[i] = 1;
                }

                for (int i = 0; i < n; i++) {
                    st.nextToken();
                    int f = (int) st.nval;
                    st.nextToken();
                    int t = (int) st.nval;
                    union(f, t);
                }
                int max = 1;
                for (int i = 1; i < 10000000; i++) {
                    max = Math.max(++num[findParent(i)], max);
                }

                System.out.println(max - 1);
            }
        }
    }

    private static void union(int f, int t) {

        int a = findParent(f);
        int b = findParent(t);
        if (a == b)
            return;
        if (a > b) {
            parent[a] = b;
        } else {
            parent[b] = a;
        }
    }

    private static int findParent(int f) {

        while (parent[f] != f) {
            f = parent[f];
        }
        return f;
    }
}
/**************************************************************
 Problem: 1444
 User: wangzhenqing
 Language: Java
 Result: Accepted
 Time:2140 ms
 Memory:105024 kb
 ****************************************************************/




